src/containers.c

```diff
@@ -0,0 +1,304 @@
+        SCFree(range);
+        range = next;
+    }
+}
+
+// base data stays in hash
+static void ContainerUrlRangeFree(void *s)
+{
+    ContainerUrlRange *cu = s;
+    SCFree(cu->key);
+    FileContainerFree(cu->files);
+    RangeContainerFree(cu);
+}
+
+static bool ContainerValueRangeTimeout(ContainerUrlRange *cu, struct timeval *ts) {
+    return (ts->tv_sec > cu->expire);
+}
+
+static void ContainerUrlRangeUpdate(ContainerUrlRange *cu, uint32_t expire) {
+    cu->expire = expire;
+}
+
+void ContainersInit(void) {
+    SCLogDebug("containers start");
+    const char *str = NULL;
+    uint64_t memcap = 0;
+    uint32_t timeout = 0;
+    if (ConfGetValue("containers.memcap", &str) == 1) {
+        if (ParseSizeStringU64(str, &memcap) < 0) {
+            SCLogWarning(SC_ERR_INVALID_VALUE,
+                         "memcap value cannot be deduced: %s,"
+                         " resetting to default",
+                         str);
+            memcap = 0;
+        }
+    }
+    if (ConfGetValue("containers.timeout", &str) == 1) {
+        if (ParseSizeStringU32(str, &timeout) < 0) {
+            SCLogWarning(SC_ERR_INVALID_VALUE,
+                         "timeout value cannot be deduced: %s,"
+                         " resetting to default",
+                         str);
+            timeout = 0;
+        }
+    }
+
+    ContainerUrlRangeList.ht = THashInit("containers.urlrange",
+                                        sizeof(ContainerUrlRange), ContainerUrlRangeSet, ContainerUrlRangeFree, ContainerUrlRangeHash,
+                                        ContainerUrlRangeCompare,
+                                        false, memcap, CONTAINER_URLRANGE_HASH_SIZE);
+    ContainerUrlRangeList.timeout = timeout;
+
+    SCLogDebug("containers started");
+}
+
+void ContainersDestroy(void) {
+    THashShutdown(ContainerUrlRangeList.ht);
+}
+
+uint32_t ContainersTimeoutHash(struct timeval *ts) {
+    uint32_t cnt = 0;
+
+    for (size_t i = 0; i < ContainerUrlRangeList.ht->config.hash_size; i++) {
+        THashHashRow *hb = &ContainerUrlRangeList.ht->array[i];
+
+        if (HRLOCK_TRYLOCK(hb) != 0)
+            continue;
+        /* hash bucket is now locked */
+        THashData *h = hb->head;
+        while (h) {
+            THashData *n = h->next;
+            if (ContainerValueRangeTimeout(h->data, ts)) {
+                /* remove from the hash */
+                if (h->prev != NULL)
+                    h->prev->next = h->next;
+                if (h->next != NULL)
+                    h->next->prev = h->prev;
+                if (hb->head == h)
+                    hb->head = h->next;
+                if (hb->tail == h)
+                    hb->tail = h->prev;
+                h->next = NULL;
+                h->prev = NULL;
+                ContainerUrlRange *c = h->data;
+                FileCloseFile(c->files, NULL, 0, FILE_TRUNCATED);
+                //TODOask can we log it somehow ?
```

Comment:
Should we log an incomplete (because of timeout) reassembled file ?
If yes, how can we get it logged from here ?
In the normal case, we transfer ownership of the `File` structure to the HTTP state